kafka-client/src/main/java/dev/responsive/kafka/store/ResponsiveRestoreListener.java

```diff
@@ -0,0 +1,102 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package dev.responsive.kafka.store;
+
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Note: this is a global object which is shared across all StreamThreads and state stores in
+ * this Streams application.
+ */
+public class ResponsiveRestoreListener implements StateRestoreListener {
+  private static final Logger LOG = LoggerFactory.getLogger(ResponsiveRestoreListener.class);
+
+  private final Metrics metrics;
+
+  private StateRestoreListener userListener;
+
+  public ResponsiveRestoreListener(final Metrics metrics) {
+    this.metrics = metrics;
+  }
+
+  public void registerUserRestoreListener(final StateRestoreListener restoreListener) {
+    userListener = restoreListener;
+  }
+
+  public StateRestoreListener userListener() {
+    return userListener;
+  }
+
+  @Override
+  public synchronized void onRestoreStart(
+      final TopicPartition topicPartition,
+      final String storeName,
+      final long startingOffset,
+      final long endingOffset
+  ) {
+    if (userListener != null) {
+      userListener.onRestoreStart(
+          topicPartition,
+          storeName,
+          startingOffset,
+          endingOffset);
+    } else {
+      LOG.info("Beginning restoration from offset {} to {} for partition {} of state store {}",
+               startingOffset, endingOffset, topicPartition, storeName);
+    }
+  }
+
+  @Override
+  public void onBatchRestored(
+      final TopicPartition topicPartition,
+      final String storeName,
+      final long batchEndOffset,
+      final long numRestored
+  ) {
+    if (userListener != null) {
+      userListener.onBatchRestored(
+          topicPartition,
+          storeName,
+          batchEndOffset,
+          numRestored);
+    } else {
+      LOG.trace("Restored {} more records up to offset {} for partition {} of state store {}",
+                numRestored, batchEndOffset, topicPartition, storeName);
+    }
+  }
+
+  @Override
+  public void onRestoreEnd(
+      final TopicPartition topicPartition,
+      final String storeName,
+      final long totalRestored
+  ) {
+    if (userListener != null) {
+      userListener.onRestoreEnd(
+          topicPartition,
+          storeName,
+          totalRestored);
+    } else {
+      LOG.info("Finished restoration of {} total records for partition {} of state store {}",
```

Comment:
why are these log lines in the `else` clause?

Reply:
I was just thinking that any/every time I've seen a user implement their own listener, it was pretty much to log exactly this.

Although there's no harm to logging something twice just to be sure. Plus, if/when we set up our own logging pipeline we'll want to make sure to get this...

I'll remove it from the else